README
NAME
    AnyEvent::Redis - Non-blocking Redis client

SYNOPSIS
      use AnyEvent::Redis;

      my $redis = AnyEvent::Redis->new(
          host => '127.0.0.1',
          port => 6379,
          encoding => 'utf8',
          on_error => sub { warn @_ },
          on_cleanup => sub { warn "Connection closed: @_" },
      );

      # callback based
      $redis->set( 'foo'=> 'bar', sub { warn "SET!" } );
      $redis->get( 'foo', sub { my $value = shift } );

      my ($key, $value) = ('list_key', 123);
      $redis->lpush( $key, $value );
      $redis->lpop( $key, sub { my $value = shift });

      # condvar based
      my $cv = $redis->lpop( $key );
      $cv->cb(sub { my $value = $_[0]->recv });

DESCRIPTION
    AnyEvent::Redis is a non-blocking (event-driven) Redis client.

    This module is an AnyEvent user; you must install and use a supported
    event loop.

ESTABLISHING A CONNECTION
    To create a new connection, use the new() method with the following
    attributes:

    host => <HOSTNAME>
        Required. The hostname or literal address of the server.

    port => <PORT>
        Optional. The server port.

    encoding => <ENCODING>
        Optional. Encode and decode data (when storing and retrieving,
        respectively) according to *ENCODING* ("utf8" is recommended or see
        Encode::Supported for details on possible *ENCODING* values).

        Omit if you intend to handle raw binary data with this connection.

    on_error => $cb->($errmsg)
        Optional. Callback that will be fired if a connection or
        database-level error occurs. The error message will be passed to the
        callback as the sole argument.

    on_cleanup => $cb->($errmsg)
        Optional. Callback that will be fired if a connection error occurs.
        The error message will be passed to the callback as the sole
        argument. After this callback, errors will be reported for all
        outstanding requests.

METHODS
    All methods supported by your version of Redis should be supported.

  Normal commands
    There are two alternative approaches for handling results from commands:

    *   AnyEvent::CondVar based:

          my $cv = $redis->command(
            # arguments to command
          );

          # Then...
          my $res;
          eval { 
              # Could die()
              $res = $cv->recv;
          }; 
          warn $@ if $@;

          # or...
          $cv->cb(sub {
            my ($cv) = @_;
            my ($result, $err) = $cv->recv
          });

    *   Callback:

          $redis->command(
            # arguments,
            sub {
              my ($result, $err) = @_;
            });

        (Callback is a wrapper around the $cv approach.)

  Transactions (MULTI/EXEC)
    Redis transactions begin with a "multi" command and end with an "exec"
    command. Commands in between are not executed immediately when they're
    sent. On receipt of the "exec", the server executes all the saved
    commands atomically, and returns all their results as one bulk reply.

    After a transaction is finished, results for each individual command are
    reported in the usual way. Thus, by the time any of these callbacks is
    called, the entire transaction is finished for better or worse.

    Results of the "exec" (containing all the other results) will be
    returned as an array reference containing all of the individual results.
    This may in some cases make callbacks on the individual commands
    unnecessary, or vice versa. In this bulk reply, errors reported for each
    individual command are represented by objects of class
    "AnyEvent::Redis::Error", which will respond to a "->message" method
    call with that error message.

    It is not permitted to nest transactions. This module does not permit
    subscription-related commands in a transaction.

  Subscriptions
    The subscription methods ("subscribe" and "psubscribe") must be used
    with a callback:

      my $cv = $redis->subscribe("test", sub {
        my ($message, $channel[, $actual_channel]) = @_;
        # ($actual_channel is provided for pattern subscriptions.)
      });

    The $cv condition will be met on unsubscribing from the channel.

    Due to limitations of the Redis protocol the only valid commands on a
    connection with an active subscription are subscribe and unsubscribe
    commands.

  Common methods
    *   get

    *   set

    *   hset

    *   hget

    *   lpush

    *   lpop

    The Redis command reference (<http://redis.io/commands>) lists all
    commands Redis supports.

REQUIREMENTS
    This requires Redis >= 1.2.

COPYRIGHT
    Tatsuhiko Miyagawa <miyagawa@bulknews.net> 2009-

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHORS
    Tatsuhiko Miyagawa

    David Leadbeater

    Chia-liang Kao

    franck cuny

    Lee Aylward

    Joshua Barratt

    Jeremy Zawodny

    Leon Brocard

    Michael S. Fischer

    Chip Salzenberg

SEE ALSO
    Redis, AnyEvent



Changes
Revision history for Perl extension AnyEvent::Redis

0.24  Wed 12 Dec 16:18:42 GMT 2012
         - Many error handling improvements (chip)
         - Reconnect logic (leedo)
         - Fix for tests with redis-server 2.6 (typester)
         - Fix leak of connections (dgl)

0.23  Wed 12 Jan 2011 23:34:58 GMT
         - Fix to always allow condvar as final param (chip)

0.23_01  Wed Dec 15 20:17:42 PST 2010
         - Added character encoding support (mfischer, dgl)
	 - New read handler, about 10% faster (mfischer)
	 - Moved read handler to AnyEvent::Redis::Protocol (mfischer)
	 - Updated documentation (mfischer)

0.22  Sun Oct 31 20:38:52 PDT 2010
        - Fixed memory leaks with MULTI commands (dgl)

0.21  Fri Oct 29 15:18:52 PDT 2010
        - Fixed async buffering issue with LRANGE command (leedo, dgl)
        - Fixed circular reference memory leaks (jzawodny)

0.20  Sun Sep 12 12:46:36 PDT 2010
        - cleanup on connection failure (leedo)

0.19_01 Wed Aug 25 12:46:59 PDT 2010
        - Added PubSub support (dgl)
        - Documentations improved (dgl)
        - Fixed KEYS for 1.2 server (dgl)
        - Bumped Test::TCP dep (dgl)

0.12  Sun Aug  8 23:34:59 PDT 2010
        - Added HSETNX (clkao)
        - Reset the state when EOF or on error (clkao)

0.11  Fri Aug  6 23:09:41 PDT 2010
        - Refactored the decoder so nested MULTI/EXEC responses work (clkao)

0.10  Fri Jul  2 17:02:16 PDT 2010
        - Support command operations on hashes for Redis 2.0 (franck)
        - Fixed MGET on non-existent keys (franck)

0.09  Thu Apr 15 10:46:46 PDT 2010
        - Fixed the test (hanekomu)

0.08  Tue Mar  9 11:44:33 JST 2010
        - Immediately send [] if the length is 0 for multi-bulk replies (leedo)

0.07  Tue Feb 23 15:40:02 PST 2010
        - Bug fix to handle nonexistent keys in multi-bulk replies (jbarratt)

0.06  Mon Feb 15 19:20:50 PST 2010
        - Add more bulk commands: getset, smove, zadd, zrem, zscore, zincrby, append (Leon Brocard)

0.05  Fri Feb 12 10:33:44 PST 2010
        - Fixed a typo in the on_error callback (jzawodn)

0.04  Tue Nov 10 22:55:59 PST 2009
        - Fixed deps

0.03  Tue Nov 10 20:33:38 PST 2009
        - Changed the object construction API to always use new()
        - Implemented on_error for callback based error handling

0.02  Tue Nov 10 17:49:39 PST 2009
        - Fixed all_cv interface
        - Fixed redis-server check in tests

0.01  Sun Nov  8 00:51:03 2009
        - original version


t/Redis.pm
package t::Redis;
use strict;
use Test::TCP;
use Test::More;
use AnyEvent::Redis;
use FindBin;

use base qw(Exporter);
our @EXPORT = qw(test_redis);

sub test_redis(&;$) {
    my $cb = shift;
    my $args = shift;

    chomp(my $redis_server = `which redis-server`);
    unless ($redis_server && -e $redis_server && -x _) {
        plan skip_all => 'redis-server not found in your PATH';
    }

    test_tcp
        server => sub {
            my $port = shift;
            rewrite_redis_conf($port);
            exec "redis-server", "t/redis.conf";
        },
        client => sub {
            my $port = shift;
            my $r = AnyEvent::Redis->new(
                host => "127.0.0.1",
                port => $port,
                ref $args ? %$args : ()
            );
            $cb->($r, $port);
        };
}

sub rewrite_redis_conf {
    my $port = shift;
    my $dir  = $FindBin::Bin;

    open my $in, "<", "t/redis.conf.base" or die $!;
    open my $out, ">", "t/redis.conf" or die $!;

    while (<$in>) {
        s/__PORT__/$port/;
        s/__DIR__/$dir/;
        print $out $_;
    }
}

END { unlink $_ for "t/redis.conf", "t/dump.rdb" }

1;



lib/AnyEvent/Redis/Protocol.pm
package AnyEvent::Redis::Protocol;
use strict;
use warnings;

=head1 NAME

AnyEvent::Redis::Protocol - Redis response adapter (read handler) for AnyEvent

=head1 DESCRIPTION

This package should not be directly used.  It provides an AnyEvent read handler
capable of parsing Redis responses using L<Protocol::Redis> underneath.

=head1 SEE ALSO

L<AnyEvent::Handle>, L<Protocol::Redis>,
L<Redis Protocol Specification|http://redis.io/topics/protocol>

=cut

sub anyevent_read_type {
    my ($handle, $cb, $p_r) = @_;
    my $rbuf = \$handle->{rbuf};

    return sub {
        return unless defined $$rbuf;

        my $input = substr $$rbuf, 0, length($$rbuf), "";
        $p_r->parse($input) if length $input;

        if(my $message = $p_r->get_message) {
            my $is_error = $message->{type} eq '-';
            $cb->(_remove_type($handle, $message->{data}), $is_error);
            return 1;
        }

        return;
    }
}

# Adapt from the Protocol::Redis format of nested hashrefs to AE::Redis's
# nested list format with a special class for errors.
sub _remove_type {
    my ($handle) = @_;

    if (ref $_[1] eq 'ARRAY') {
        [map +($_->{type} eq '-'
                      ? bless \$_->{data}, "AnyEvent::Redis::Error"
                      : _remove_type($handle, $_->{data})
                ), @{$_[1]}];
    }
    elsif (ref $_[1] eq 'HASH') {
        _remove_type($handle, $_[1]->{data});
    }
    else {
        ($handle->{encoding} && length $_[1])
          ? $handle->{encoding}->decode($_[1])
          : $_[1];
    }
}

=head1 AUTHOR

Michael S. Fischer <michael+cpan@dynamine.net>

David Leadbeater <dglE<xFE6B>dgl.cx>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2010 Michael S. Fischer.

Copyright (C) 2011 David Leadbeater.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;

__END__

# ex:sw=4:et:


lib/AnyEvent/Redis.pm
package AnyEvent::Redis;

use strict;
use 5.008_001;
our $VERSION = '0.24';

use constant DEBUG => $ENV{ANYEVENT_REDIS_DEBUG};
use AnyEvent;
use AnyEvent::Handle;
use AnyEvent::Socket;
use AnyEvent::Redis::Protocol;
use Carp qw( croak confess );
use Encode ();
use Scalar::Util qw(weaken);

our $AUTOLOAD;

sub new {
    my ($class, %args) = @_;

    my $host = delete $args{host} || '127.0.0.1';
    my $port = delete $args{port} || 6379;

    if (my $encoding = $args{encoding}) {
        $args{encoding} = Encode::find_encoding($encoding);
        croak qq{Encoding "$encoding" not found} unless ref $args{encoding};
    }

    bless {
        host => $host,
        port => $port,
        pending_cvs => [],
        %args,
    }, $class;
}

sub run_cmd {
    my $self = shift;
    my $cmd  = shift;

    $self->{cmd_cb} or return $self->connect($cmd, @_);
    $self->{cmd_cb}->($cmd, @_);
}

sub DESTROY { }

sub AUTOLOAD {
    my $self = shift;
    (my $method = $AUTOLOAD) =~ s/.*:://;
    $self->run_cmd($method, @_);
}

sub all_cv {
    my $self = shift;
    $self->{all_cv} = shift if @_;
    $self->{all_cv} ||= AE::cv;
}

sub cleanup {
    my $self = shift;
    delete $self->{cmd_cb};
    delete $self->{sock};
    $self->{on_error}->(@_) if $self->{on_error};
    $self->{on_cleanup}->(@_) if $self->{on_cleanup};
    for (splice(@{$self->{pending_cvs}}),
         splice(@{$self->{multi_cvs} || []}))
    {
        eval { $_->croak(@_) };
        warn "Exception in cleanup callback (ignored): $@" if $@;
    }
    return;
}

sub connect {
    my $self = shift;

    my $cv;
    if (@_) {
        $cv = pop if UNIVERSAL::isa($_[-1], 'AnyEvent::CondVar');
        $cv ||= AE::cv;
        push @{$self->{connect_queue}}, [ $cv, @_ ];
    }

    return $cv if $self->{sock};
    weaken $self;

    $self->{sock} = tcp_connect $self->{host}, $self->{port}, sub {
        my $fh = shift
            or do {
              my $err = "Can't connect Redis server: $!";
              $self->cleanup($err);
              eval { $cv->croak($err) };
              warn "Exception in connect failure callback (ignored): $@" if $@;
              return
            };

        binmode $fh; # ensure bytes until we decode

        my $hd = AnyEvent::Handle->new(
            fh => $fh,
            on_error => sub { $_[0]->destroy;
                              $self->cleanup($_[2]) if $_[1];
                          },
            on_eof   => sub { $_[0]->destroy;
                              $self->cleanup('connection closed');
                          },
            encoding => $self->{encoding},
        );

        $self->{cmd_cb} = sub {
            my $command = lc shift;
            my $is_pubsub    = $command =~ /^p?(?:un)?subscribe\z/;
            my $is_subscribe = $command =~ /^p?subscribe\z/;


            # Are we already subscribed to anything?
            if ($self->{sub} && %{$self->{sub}}) {
                croak "Use of non-pubsub command during pubsub session may result in unexpected behaviour"
                  unless $is_pubsub;
            }
            # Are we already in a transaction?
            if ($self->{multi_write}) {
                croak "Use of pubsub or multi command in transaction is not supported"
                  if $is_pubsub || $command eq 'multi';
            } else {
                croak "Can't 'exec' a transaction because none is pending"
                  if $command eq 'exec';
            }

            my ($cv, $cb);
            if (@_) {
                $cv = pop if ref $_[-1] && UNIVERSAL::isa($_[-1], 'AnyEvent::CondVar');
                $cb = pop if ref $_[-1] eq 'CODE';
            }
            $cv ||= AE::cv;
            croak "Must provide a CODE reference for subscriptions" if $is_subscribe && !$cb;

            my $send = join("\r\n",
                  "*" . (1 + @_),
                  map { ('$' . length $_ => $_) }
                        (uc($command), map { $self->{encoding} && length($_)
                                             ? $self->{encoding}->encode($_)
                                             : $_ } @_))
                . "\r\n";

            warn $send if DEBUG;

            # $self is weakened to avoid leaks, hold on to a strong copy
            # controlled via a CV.
            my $cmd_cv = AE::cv;
            $cmd_cv->cb(sub {
                my $strong_self = $self;
              });

            # pubsub is very different - get it out of the way first

            if ($is_pubsub) {

                $hd->push_write($send);

                my $already = $self->{sub} && %{$self->{sub}};

                if ($is_subscribe) {
                    $self->{sub}->{$_} ||= [$cv, $cb] for @_;
                }

                if (!$already && @_) {
                    my $res_cb; $res_cb = sub {
                        $hd->push_read("AnyEvent::Redis::Protocol" => sub {
                            my ($res, $err) = @_;

                            if (ref $res) {
                                my $action = lc $res->[0];
                                warn "$action $res->[1]" if DEBUG;

                                if ($action eq 'message') {
                                    $self->{sub}->{$res->[1]}[1]->($res->[2], $res->[1]);

                                } elsif ($action eq 'pmessage') {
                                    $self->{sub}->{$res->[1]}[1]->($res->[3], $res->[2], $res->[1]);

                                } elsif ($action eq 'subscribe' || $action eq 'psubscribe') {
                                    $self->{sub_count} = $res->[2];

                                } elsif ($action eq 'unsubscribe' || $action eq 'punsubscribe') {
                                    $self->{sub_count} = $res->[2];
                                    eval { $self->{sub}->{$res->[1]}[0]->send };
                                    warn "Exception in callback (ignored): $@" if $@;
                                    delete $self->{sub}->{$res->[1]};
                                    $self->all_cv->end;
                                    $cmd_cv->send;

                                } else {
                                    warn "Unknown pubsub action: $action";
                                }
                            }

                            if ($self->{sub_count} || %{$self->{sub}}) {
                                # Carry on reading while we are subscribed
                                $res_cb->();
                            }
                        });
                    };

                    $res_cb->();
                }

                return $cv;
            }

            # non-pubsub from here on out

            $cv->cb(sub {
                my $cv = shift;
                local $@;
                eval {
                    my $res = $cv->recv;
                    $cb->($res);
                };
                if ($@) {
                    ($self->{on_error} || sub { die @_ })->(my $err = $@);
                }
            }) if $cb;

            $self->all_cv->begin;
            push @{$self->{pending_cvs}}, $cv;

            $hd->push_write($send);

            if ($command eq 'exec') {

                # at end of transaction, expect bulk reply possibly including errors
                $hd->push_read("AnyEvent::Redis::Protocol" => sub {
                    my ($res, $err) = @_;

                    $self->_expect($cv);

                    my @mcvs = splice @{$self->{multi_cvs}};

                    if ($err || ref($res) ne 'ARRAY') {
                        for ($cv, @mcvs) {
                            eval { $_->croak($res, 1) };
                            warn "Exception in callback (ignored): $@" if $@;
                        }
                    } else {
                        for my $i (0 .. $#mcvs) {
                            my $r = $res->[$i];
                            eval {
                                ref($r) && UNIVERSAL::isa($r, 'AnyEvent::Redis::Error')
                                  ? $mcvs[$i]->croak($$r)
                                  : $mcvs[$i]->send($r);
                            };
                            warn "Exception in callback (ignored): $@" if $@;
                        }
                        eval { $cv->send($res) };
                        warn "Exception in callback (ignored): $@" if $@;
                    }

                    $self->all_cv->end;
                    $cmd_cv->send;
                });

                delete $self->{multi_write};

            } elsif ($self->{multi_write}) {

                # in transaction, expect only "QUEUED"
                $hd->push_read("AnyEvent::Redis::Protocol" => sub {
                    my ($res, $err) = @_;

                    $self->_expect($cv);

                    if (!$err && $res eq 'QUEUED') {
                        push @{$self->{multi_cvs}}, $cv;
                    }
                    else {
                        eval { $cv->croak($res) };
                        warn "Exception in callback (ignored): $@" if $@;
                    }

                    $self->all_cv->end;
                    $cmd_cv->send;
                });

            } else {

                $hd->push_read("AnyEvent::Redis::Protocol" => sub {
                    my ($res, $err) = @_;

                    $self->_expect($cv);

                    if ($command eq 'info') {
                        $res = { map { split /:/, $_, 2 } grep !/^#/, split /\r\n/, $res };
                    } elsif ($command eq 'keys' && !ref $res) {
                        # Older versions of Redis (1.2) need this
                        $res = [split / /, $res];
                    }

                    eval { $err ? $cv->croak($res) : $cv->send($res) };
                    warn "Exception in callback (ignored): $@" if $@;

                    $self->all_cv->end;
                    $cmd_cv->send;
                });

                $self->{multi_write} = 1 if $command eq 'multi';

            }

            return $cv;
        };

        my $queue = delete $self->{connect_queue} || [];
        for my $command (@$queue) {
            my($cv, @args) = @$command;
            $self->{cmd_cb}->(@args, $cv);
        }

    };

    return $cv;
}

sub _expect {
    my ($self, $cv) = @_;
    my $p = shift @{$self->{pending_cvs} || []};
    $p && $p == $cv or confess "BUG: mismatched CVs";
}

1;
__END__

=encoding utf-8

=for stopwords

=head1 NAME

AnyEvent::Redis - Non-blocking Redis client

=head1 SYNOPSIS

  use AnyEvent::Redis;

  my $redis = AnyEvent::Redis->new(
      host => '127.0.0.1',
      port => 6379,
      encoding => 'utf8',
      on_error => sub { warn @_ },
      on_cleanup => sub { warn "Connection closed: @_" },
  );

  # callback based
  $redis->set( 'foo'=> 'bar', sub { warn "SET!" } );
  $redis->get( 'foo', sub { my $value = shift } );

  my ($key, $value) = ('list_key', 123);
  $redis->lpush( $key, $value );
  $redis->lpop( $key, sub { my $value = shift });

  # condvar based
  my $cv = $redis->lpop( $key );
  $cv->cb(sub { my $value = $_[0]->recv });

=head1 DESCRIPTION

AnyEvent::Redis is a non-blocking (event-driven) Redis client.

This module is an AnyEvent user; you must install and use a supported event loop.

=head1 ESTABLISHING A CONNECTION

To create a new connection, use the new() method with the following attributes:

=over

=item host => <HOSTNAME>

B<Required.>  The hostname or literal address of the server.  

=item port => <PORT>

Optional.  The server port.

=item encoding => <ENCODING>

Optional.  Encode and decode data (when storing and retrieving, respectively)
according to I<ENCODING> (C<"utf8"> is recommended or see L<Encode::Supported>
for details on possible I<ENCODING> values).

Omit if you intend to handle raw binary data with this connection.

=item on_error => $cb->($errmsg)

Optional.  Callback that will be fired if a connection or database-level error
occurs.  The error message will be passed to the callback as the sole argument.

=item on_cleanup => $cb->($errmsg)

Optional.  Callback that will be fired if a connection error occurs.  The
error message will be passed to the callback as the sole argument.  After
this callback, errors will be reported for all outstanding requests.

=back

=head1 METHODS

All methods supported by your version of Redis should be supported.

=head2 Normal commands

There are two alternative approaches for handling results from commands:

=over 4

=item * L<AnyEvent::CondVar> based:

  my $cv = $redis->command(
    # arguments to command
  );

  # Then...
  my $res;
  eval { 
      # Could die()
      $res = $cv->recv;
  }; 
  warn $@ if $@;

  # or...
  $cv->cb(sub {
    my ($cv) = @_;
    my ($result, $err) = $cv->recv
  });


=item * Callback:

  $redis->command(
    # arguments,
    sub {
      my ($result, $err) = @_;
    });

(Callback is a wrapper around the C<$cv> approach.)

=back

=head2 Transactions (MULTI/EXEC)

Redis transactions begin with a "multi" command and end with an "exec"
command.  Commands in between are not executed immediately when they're
sent.  On receipt of the "exec", the server executes all the saved commands
atomically, and returns all their results as one bulk reply.

After a transaction is finished, results for each individual command are
reported in the usual way.  Thus, by the time any of these callbacks is
called, the entire transaction is finished for better or worse.

Results of the "exec" (containing all the other results) will be returned as
an array reference containing all of the individual results.  This may in
some cases make callbacks on the individual commands unnecessary, or vice
versa.  In this bulk reply, errors reported for each individual command are
represented by objects of class C<AnyEvent::Redis::Error>, which will
respond to a C<< ->message >> method call with that error message.

It is not permitted to nest transactions.  This module does not permit
subscription-related commands in a transaction.

=head2 Subscriptions

The subscription methods (C<subscribe> and C<psubscribe>) must be used with a callback:

  my $cv = $redis->subscribe("test", sub {
    my ($message, $channel[, $actual_channel]) = @_;
    # ($actual_channel is provided for pattern subscriptions.)
  });

The C<$cv> condition will be met on unsubscribing from the channel.

Due to limitations of the Redis protocol the only valid commands on a
connection with an active subscription are subscribe and unsubscribe commands.

=head2 Common methods

=over 4

=item * get

=item * set

=item * hset

=item * hget

=item * lpush

=item * lpop

=back

The Redis command reference (L<http://redis.io/commands>) lists all commands
Redis supports.

=head1 REQUIREMENTS

This requires Redis >= 1.2.

=head1 COPYRIGHT

Tatsuhiko Miyagawa E<lt>miyagawa@bulknews.netE<gt> 2009-

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHORS

Tatsuhiko Miyagawa

David Leadbeater

Chia-liang Kao

franck cuny

Lee Aylward

Joshua Barratt

Jeremy Zawodny

Leon Brocard

Michael S. Fischer

Chip Salzenberg

=head1 SEE ALSO

L<Redis>, L<AnyEvent>

=cut
